### ISSUE TITLE
Components: Order of elements should correlate with definition, even with sub-components

### ISSUE THREAD

**Issue by mawid6** [plantuml/plantuml#1018]
Hi!
### **TLDR:**
* ✔️ If I define a component diagram `component A, B, C` etc, the resulting diagram correlates with the specified order (i  e how I entered it). 
* ❌ If I define a similar component diagram, but include a breakdown of `A` into `A.1` and `A.2` - the order of the components no longer correlates with the specified order.

_Not discussing if this is bug/by design, etc. Just raising this._

## ✔️ Order correlates for simple list

```
@startuml
component A 
component B
component C
component D
@enduml
``` 
![image](https://user-images.githubusercontent.com/77722885/168787145-3c584bf0-ff18-4df8-aeab-92efd2934f53.png)

## ❌ Order does not correlate as soon as you add sub-components

```
@startuml
component A {
component A.1
}
component B
component C
component D
@enduml
``` 
### Expected:
Diagram correlates with entered order, i e:
![image](https://user-images.githubusercontent.com/77722885/168788848-fa0565c6-2fba-47cf-9d59-496a2ed64893.png)

### Actual:
![image](https://user-images.githubusercontent.com/77722885/168787447-cec2f395-d9eb-40d4-9de5-d69d13425f4e.png)

### Diagram link
[Code on plantuml server](http://www.plantuml.com/plantuml/umla/SoWkIImgAStDuKhEpot8pqlDAr5mLAX6vkaPSjKYyPsGsCv8R1SkXzIy5A190000), currently yields: 
![generated diagram](http://www.plantuml.com/plantuml/png/SoWkIImgAStDuKhEpot8pqlDAr5mLAX6vkaPSjKYyPsGsCv8R1SkXzIy5A190000)

----
# Target audience
* Any casual dev  / BA / SA that wants to quickly get a component diagram or other similar boxes diagram down.
* Code generating boxes diagrams from code or config, without having the opportunity to add layouting control instructions in order to make it correlate with the order.
 
# Out of scope
* Whether this is by design or a bug or whatever. Raising this from a UX POV, what is a good user experience, is this not a reasonable expectation to fulfill...? 😃 
* Another area of reasonable expectation might be around when shapes are populated in a new row.

# The Common Workaround: layout controlling instructions
In order to "coerce"/get a certain layout, different additional instructions can be added (e g the by now standard `-[hidden]-`, `-left-`, etc, etc) - but please note:

> ℹ️ This issue is not a cry for help about achieving a certain layout, but rather raising this rather (on the surface) simple and reasonable expectation to increase the connection between syntax and intent and the contents of the generated image.

# Discussion / notes / references
* In general, graphviz is of course an enabler and constrainer w r t layout expectations.
* I think this expectation might go a long way to start making the graph generated diagrams more intuitive and appealing to use. 
* I did search to try and find a similar issue. I found these:
  * plantuml/plantuml#13 
  * plantuml/backlog#35
* If you find this expectation of order unreasonable, it would be interesting to hear a clear argument as to why.

**Comment by mawid6** [comment]
Fantastic reply and summary of links to good info and discussions. 
Thanks alot! I will learn more.

( Thinking out loud here... I wonder if a simplified generation engine / diagram type would be a conceivable option, for scenarios where you don't want so much autolayout, and perhaps even more to the point, where a treemap is sufficient as a layout - a boxes representation of a tree, not a graph, for instance.  An example: A business capability model - boxes in boxes, a breakdown. )

**Comment by The-Lum** [comment]
Hi all,

_[This is an Issue Review]_
This is now answered.
- Thanks @mawid6 for the report.
- Thanks PlantUML team for your work.
	
Regards.

**Comment by mawid6** [comment]
To @The-Lum and anyone else reading this: It is correct that someone answered / replied in this thread. The reply included links to more info that helps with different workarounds, none of which work so well and none of which answer the actual issue I reported. As I wrote in the issue, I did not raise this to gather workarounds, I raised this to point towards the core / basic problem. Despite this, I replied politely thanking for the information - but again, none of it address the issue, but are only workarounds. So: 

The issue has certainly not been fixed. 

The issue is still there. It is a basic / core issue and I suspect it is not easy to solve. 


BR! /m
I really like plantuml, have been advocating it for over a decade, and will continue to do so.

**Comment by The-Lum** [comment]
Hi @mawid6,

>The issue has certainly not been fixed.
The issue is still there. It is a basic / core issue and I suspect it is not easy to solve.

I may have closed it too quickly
No problem to re-open this issue. Thanks @arnaudroques.

Regards.

**Comment by mawid6** [comment]
@The-Lum: Completely fine, of course! 

To be honest, I think this one may be tough to solve in a good way. Not sure, just a feeling.
But I think it may be good to have this simple test and display of some of the layouting issues documented.

Thanks again!
